### ISSUE TITLE
Softlock possibility with the new feature of paying before obtaining a scrub item

### ISSUE THREAD

**Issue by RealLarsP** [TestRunnerSRL/OoT-Randomizer#2285]
Saying "Yes" to a scrub when getting his item apparently sets the flag for the scrub item to be obtained. However, if the scrub area to speak to him is bigger than the area where you can obtain the item, you can walk away without getting the item. If you happen to reload the area, the scrub is gone and you'll never be able to get his item, even though logic thinks you already got it.

Tested and confirmed on Roman's branch, v8.1.47 R-1

**Comment by fenhl** [comment]
The example you gave on Discord was the `Deku Tree MQ Deku Scrub` location, where you can talk to the scrub from below but can't actually get the item unless you're on the ledge. One proposed fix was to use the delayed item queue for scrub items but that would have logical implications in this case, as you would then be able to get the item from below. Maybe instead, the radius for talking to the scrub can be changed to match the get item trigger?

**Comment by flagrama** [comment]
The problem is Z-Targetting ignores all of these distance values.

**Comment by fenhl** [comment]
What's the vanilla behavior here?

**Comment by RealLarsP** [comment]
In vanilla, it seems to be the same problem. I just tried it with the Deku Tree MQ scrub again. In vanilla, the scrub is infinite, but I was able to stun him from below, buy his Deku Shield and didn't get anything. If I now get onto the ledge and close enough, I get the shield and pay the 50 rupees. But if I reload the area, the scrub just resets and continues attacking me.

**Comment by djevangelia** [comment]
It's because for talking, there is only a requirement to be within 130 units XZ distance (see `EnDns_Idle`) which reaches slightly outside the ledge, but for giving the item, there is also a requirement to be within 100 units Y distance, which just barely is out of reach from the water.

Changing the last argument of `Actor_OfferGetItem(&this->actor, play, this->dnsItemEntry->getItemId, 130.0f, 100.0f)` in `EnDns_OfferSaleItem` fixes this. I just tried 130 to ensure it works and it did, but for safety's sake maybe a higher value is better.
This only changes scrubs that don't sell upgrades.

**Comment by flagrama** [comment]
I don't suppose there is a way to resolve this that retains the expected logic of being on the ledge?

**Comment by fenhl** [comment]
I'm assuming changing the second-to-last argument to `100.0` would accomplish that?

**Comment by fenhl** [comment]
Wait, I misread, I thought you were saying the horizontal radius is different.

**Comment by fenhl** [comment]
Yeah, to fix this in a way that doesn't allow accessing the item from below, we'd need to find a way to limit the Y range for talking.

**Comment by djevangelia** [comment]
Encasing the entire talk offer function in an `actor.yDistToPlayer` check should work for that, and using same distance as the get item function.
I tried in MQ Deku Tree and I can't get him to talk to me from below the ledge as child or adult.

Now, this affects all sale Deku Scrubs, but if every other scrub gives their item instantly after purchase, it shouldn't make any difference for the others.

There might be a risk of starting to talk to the scrub, getting out of get item range on the same frame, and then falling down the ledge and unable to get the item. For this reason increasing the Y distance of `Actor_OfferGetItem(&this->actor, play, this->dnsItemEntry->getItemId, 130.0f, 100.0f)` slightly might still be a good idea.

```
void EnDns_Idle(EnDns* this, PlayState* play) {
    Math_SmoothStepToS(&this->actor.shape.rot.y, this->actor.yawTowardsPlayer, 3, 2000, 0);
    this->actor.world.rot.y = this->actor.shape.rot.y;

    **if( fabsf(this->actor.yDistToPlayer) < 100.0f) {**
        if (Actor_TalkOfferAccepted(&this->actor, play) ) {
            this->actionFunc = EnDns_Talk;
        } else {
            if ((this->collider.base.ocFlags1 & OC1_HIT) || (this->actor.isLockedOn)) {
                this->actor.flags |= ACTOR_FLAG_TALK_OFFER_AUTO_ACCEPTED;
            } else {
                this->actor.flags &= ~ACTOR_FLAG_TALK_OFFER_AUTO_ACCEPTED;
            }
            if (this->actor.xzDistToPlayer < 130.0) {
                Actor_OfferTalkNearColChkInfoCylinder(&this->actor, play);
            }
        }
    }
}
```
